sbol/identified.py

```diff
@@ -189,3 +189,251 @@ def update_uri(self):
+                return uri[:rlimit]
+            rlimit = uri.rfind('/') + 1
+            if rlimit:
+                return uri[:rlimit]
+            return ''
+
+        new_obj = self.__class__()
+
+        # Assign the new object to the target Document
+        if target_doc:
+            new_obj.doc = target_doc
+
+        # This namespace map will be used later when copying namespaces over to
+        # the new Document
+        if self.doc:
+            namespace_map = {ns: p for p, ns in self.doc._namespaces.items()}
+
+        # Copy properties
+        for property_uri, value_store in self.properties.items():
+            new_obj.properties[property_uri] = value_store.copy()
+
+            # Add a non-default namespace to the target document if not present
+            # (This can happen when copying extension properties not in the
+            # SBOL namespace, for example.)
+            if self.doc and target_doc != None:
+                property_namespace = URIRef(parseNamespace(property_uri))
+                if property_namespace in namespace_map.keys():
+                    prefix = namespace_map[property_namespace]
+                    target_doc.addNamespace(property_namespace, prefix)
+
+        # If caller specified a target_namespace argument, then import objects into this new namespace
+        # This involves replacing the target_namespace in ReferenceObject URIs with the current Homespace
+        # Don't overwrite namespaces for the wasDerivedFrom field, which points back to the original object
+        if target_namespace:
+
+            # Collect ReferencedObject attributes
+            reference_properties = [p for p in new_obj.__dict__.values() if isinstance(p, ReferencedObject)]
+
+            # These URIProperty attributes should be treated like ReferencedObject attributes
+            if '_identity' in new_obj.__dict__.keys():
+                reference_properties.append(new_obj.__dict__['_identity'])
+            if '_persistentIdentity' in new_obj.__dict__.keys():
+                reference_properties.append(new_obj.__dict__['_persistentIdentity'])
+
+            for reference_property in reference_properties:
+                if reference_property._upperBound == '1':
+                    values = [reference_property.value]
+                else:
+                    values = reference_property.value
+                for i_val, val in enumerate(values):
+                    if target_namespace in val:
+                        replacement_target = target_namespace
+                        replacement = getHomespace()
+                        new_val = val.replace(replacement_target, replacement)
+                        values[i_val] = new_val
+                if reference_property._upperBound == '1':
+                    reference_property.value = values[0]
+                else:
+                    reference_property.value = values
+                print(reference_property._rdf_type, reference_property.value)
```

Comment:
Can this be extracted into a separate method so that the copy method is a little more tractable? This feels like a good candidate for extraction.

Reply:
This section is more condensed and perhaps readable now.  I factored out some of the namespace operations into `replace_namespace` method